File: kmp-workshop/composeApp/src/androidMain/kotlin/com/chmelik/kmp_workshop/BookViewModel.kt
package com.chmelik.kmp_workshop

import android.content.Context
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel

class BookViewModel(context: Context) : ViewModel() {
    private val bookService = BookService(context)

    var books by mutableStateOf(listOf<Book>())
        private set

    var likedIDs by mutableStateOf(setOf<String>())
        private set

    fun fetchBooks() {
        books = bookService.fetchBooks()
        likedIDs = books
            .filter { bookService.isBookLiked(it.id) }
            .map { it.id }
            .toSet()
    }

    fun isBookLiked(id: String): Boolean {
        return likedIDs.contains(id)
    }

    fun likeUnlikeBook(id: String) {
        if (likedIDs.contains(id)) {
            bookService.unlikeBook(id)
            likedIDs -= id
        } else {
            bookService.likeBook(id)
            likedIDs += id
        }
    }
}

File: kmp-workshop/iosApp/iosApp/BookViewModel.swift
//
// Created by Dominik Chmelík on 15.10.2025.
//

import SwiftUI
import Observation

@MainActor
@Observable
final class BookViewModel {
    @ObservationIgnored
    private let bookService: BookService = BookService()

    private(set) var books: [Book] = []
    private var likedIDs: Set<String> = []

    func fetchBooks() {
        books = bookService.fetchBooks()
        likedIDs = Set(books.filter { bookService.isBookLiked(id: $0.id) }.map { $0.id })
    }

    func isBookLiked(id: String) -> Bool {
        return likedIDs.contains(id)
    }

    func likeUnlikeBook(id: String) {
        if likedIDs.contains(id) {
            bookService.unlikeBook(id: id)
            likedIDs.remove(id)
        } else {
            bookService.likeBook(id: id)
            likedIDs.insert(id)
        }
    }
}
